File: src/Exporter/GPXExporter.cpp
#include "Stdafx.h"
#include "MyMemory.h"
#include "Data/DateTime.h"
#include "Exporter/GPXExporter.h"
#include "IO/BufferedOutputStream.h"
#include "IO/StreamWriter.h"
#include "Map/GPSTrack.h"
#include "Text/MyString.h"
#include "Text/MyStringFloat.h"

Exporter::GPXExporter::GPXExporter()
{
	this->codePage = 65001;
}

Exporter::GPXExporter::~GPXExporter()
{
}

Int32 Exporter::GPXExporter::GetName()
{
	return *(Int32*)"GPXE";
}

IO::FileExporter::SupportType Exporter::GPXExporter::IsObjectSupported(NN<IO::ParsedObject> pobj)
{
	if (pobj->GetParserType() != IO::ParserType::MapLayer)
	{
		return IO::FileExporter::SupportType::NotSupported;
	}
	NN<Map::MapDrawLayer> layer = NN<Map::MapDrawLayer>::ConvertFrom(pobj);
	if (layer->GetObjectClass() != Map::MapDrawLayer::OC_GPS_TRACK)
	{
		return IO::FileExporter::SupportType::NotSupported;
	}
	return IO::FileExporter::SupportType::NormalStream;
}

Bool Exporter::GPXExporter::GetOutputName(UOSInt index, UnsafeArray<UTF8Char> nameBuff, UnsafeArray<UTF8Char> fileNameBuff)
{
	if (index == 0)
	{
		Text::StrConcatC(nameBuff, UTF8STRC("GPX file"));
		Text::StrConcatC(fileNameBuff, UTF8STRC("*.gpx"));
		return true;
	}
	return false;
}

void Exporter::GPXExporter::SetCodePage(UInt32 codePage)
{
	this->codePage = codePage;
}

Bool Exporter::GPXExporter::ExportFile(NN<IO::SeekableStream> stm, Text::CStringNN fileName, NN<IO::ParsedObject> pobj, Optional<ParamData> param)
{
	if (pobj->GetParserType() != IO::ParserType::MapLayer)
	{
		return false;
	}
	NN<Map::MapDrawLayer> layer = NN<Map::MapDrawLayer>::ConvertFrom(pobj);
	if (layer->GetObjectClass() != Map::MapDrawLayer::OC_GPS_TRACK)
		return false;
	NN<Map::GPSTrack> track = NN<Map::GPSTrack>::ConvertFrom(layer);
	UTF8Char sbuff[256];
	UnsafeArray<UTF8Char> sptr;

	UOSInt i;
	UOSInt j;
	UOSInt k;
	UOSInt l;
	UnsafeArray<Map::GPSTrack::GPSRecord3> recs;
	Data::DateTime dt;

	Text::Encoding enc(this->codePage);
	IO::BufferedOutputStream cstm(stm, 65536);
	IO::StreamWriter writer(cstm, &enc);

	writer.Write(CSTR("<?xml version=\"1.0\" encoding=\""));
	sptr = Text::EncodingFactory::GetInternetName(sbuff, this->codePage);
	writer.Write(CSTRP(sbuff, sptr));
	writer.WriteLine(CSTR("\"?>"));
	writer.WriteLine(CSTR("<gpx version=\"v1.0\" creator=\"iTravel Tech Inc. - http://www.itravel-tech.com\">"));
	writer.WriteLine(CSTR("<trk>"));
	writer.WriteLine(CSTR("<name>Track</name>"));
	i = 0;
	j = track->GetTrackCnt();
	while (i < j)
	{
		if (track->GetTrack(i, l).SetTo(recs))
		{
			writer.WriteLine(CSTR("<trkseg>"));
			k = 0;
			while (k < l)
			{
				writer.Write(CSTR("<trkpt lat=\""));
				sptr = Text::StrDouble(sbuff, recs[k].pos.GetLat());
				writer.Write(CSTRP(sbuff, sptr));
				writer.Write(CSTR("\" lon=\""));
				sptr = Text::StrDouble(sbuff, recs[k].pos.GetLon());
				writer.Write(CSTRP(sbuff, sptr));
				writer.WriteLine(CSTR("\">"));
				
				writer.Write(CSTR("<ele>"));
				sptr = Text::StrDouble(sbuff, recs[k].altitude);
				writer.Write(CSTRP(sbuff, sptr));
				writer.WriteLine(CSTR("</ele>"));

				writer.Write(CSTR("<time>"));
				dt.SetInstant(recs[k].recTime);
				sptr = dt.ToString(sbuff, "yyyy-MM-ddTHH:mm:ssZ");
				writer.Write(CSTRP(sbuff, sptr));
				writer.WriteLine(CSTR("</time>"));

				writer.Write(CSTR("<desc>lat.="));
				sptr = Text::StrDoubleFmt(sbuff, recs[k].pos.GetLat(), "0.000000");
				writer.Write(CSTRP(sbuff, sptr));
				writer.Write(CSTR(", lon.="));
				sptr = Text::StrDoubleFmt(sbuff, recs[k].pos.GetLon(), "0.000000");
				writer.Write(CSTRP(sbuff, sptr));
				writer.Write(CSTR(", Alt.="));
				sptr = Text::StrDoubleFmt(sbuff, recs[k].altitude, "0.000000");
				writer.Write(CSTRP(sbuff, sptr));
				writer.Write(CSTR("m, Speed="));
				sptr = Text::StrDoubleFmt(sbuff, recs[k].speed * 1.852, "0.000000");
				writer.Write(CSTRP(sbuff, sptr));
				writer.WriteLine(CSTR("m/h.</desc>"));

				writer.Write(CSTR("<speed>"));
				sptr = Text::StrDoubleFmt(sbuff, recs[k].speed * 1.852 / 3.6, "0.000000");
				writer.Write(CSTRP(sbuff, sptr));
				writer.WriteLine(CSTR("</speed>"));

				writer.WriteLine(CSTR("</trkpt>"));

				k++;
			}
			writer.WriteLine(CSTR("</trkseg>"));
		}
		i++;
	}

	writer.WriteLine(CSTR("</trk>"));
	writer.WriteLine(CSTR("</gpx>"));
	return true;
}

File: js/@sswroom/sswr/exporter/GPXExporter.js
import * as data from "../data.js";
import * as map from "../map.js";
import * as math from "../math.js";

export class GPXExporter
{
	constructor()
	{
	}

	getName()
	{
		return "GPSExporter";
	}

	/**
	 * @param {data.ParsedObject} pobj
	 */
	isObjectSupported(pobj)
	{
		if (pobj instanceof map.GPSTrack)
		{
			return true;
		}
		return false;
	}

	getOutputExt()
	{
		return "gpx";
	}

	getOutputMIME()
	{
		return "application/gpx+xml";
	}

	/**
	 * @param {string} fileName
	 * @param {data.ParsedObject} pobj
	 * @param {object|undefined} param
	 */
	exportFile(fileName, pobj, param)
	{
		let track = pobj;
		if (!(track instanceof map.GPSTrack))
			return null;
		let sb = [];
		let i;
		let j;
		let k;
		let l;
		let recs;
	
		sb.push("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\r\n");
		sb.push("<gpx version=\"v1.0\" creator=\"sswr - https://github.com/sswroom/SClass\">\r\n");
		sb.push("<trk>\r\n");
		sb.push("<name>Track</name>\r\n");
		i = 0;
		j = track.getTrackCnt();
		while (i < j)
		{
			if ((recs = track.getTrack(i)) != null)
			{
				sb.push("<trkseg>\r\n");
				k = 0;
				l = recs.length;
				while (k < l)
				{
					sb.push("<trkpt lat=\""+recs[k].lat+"\" lon=\""+recs[k].lon+"\">\r\n");
					sb.push("<ele>"+recs[k].altitude+"</ele>");
					sb.push("<time>"+data.Timestamp.fromTicks(recs[k].recTime, 0).toString("yyyy-MM-ddTHH:mm:ssZ")+"</time>\r\n");
					sb.push("<desc>lat.="+math.roundToStr(recs[k].lat, 6)+", lon.="+math.roundToStr(recs[k].lon, 6));
					sb.push(", Alt.="+math.roundToStr(recs[k].altitude, 6)+"m, Speed="+math.roundToStr(recs[k].speed * 1.852, 6)+"m/h.</desc>\r\n");
					sb.push("<speed>"+math.roundToStr(recs[k].speed * 1.852 / 3.6, 6)+"</speed>\r\n");
					sb.push("</trkpt>\r\n");
	
					k++;
				}
				sb.push("</trkseg>\r\n");
			}
			i++;
		}
	
		sb.push("</trk>\r\n");
		sb.push("</gpx>\r\n");
		return new TextEncoder().encode(sb.join(""));
	}
};
